lib/algo/manifold_interp/SnapshotInterpolator.cpp

```diff
@@ -0,0 +1,222 @@
+    CAROM_VERIFY(snapshot_ts.size() > 0);
+    CAROM_VERIFY(output_ts.size() > 0);
+    CAROM_VERIFY(output_snapshots.size() == 0);
+    CAROM_VERIFY(snapshot_ts[0]->getData()[0] - FLT_EPSILON <=
+                 output_ts[0]->getData()[0]  &&
+                 output_ts[output_ts.size()-1]->getData()[0] <= snapshot_ts[snapshot_ts.size()
+                         -1]->getData()[0] + FLT_EPSILON);
+
+    int n_out = output_ts.size();
+    int n_snap = snapshots.size();
+    int n_dim = snapshots[0]->dim();
+
+    for(int i = 0; i < n_out; ++i)
+    {
+        Vector* temp_snapshot = new Vector(snapshots[0]->dim(),
+                                           snapshots[0]->distributed());
+        output_snapshots.push_back(temp_snapshot);
+    }
+
+    for(int i = 0; i < n_dim; ++i)
+    {
+        double h_temp,delta_temp, t;
+        std::vector<double> h,d,delta,t_in,y_in;
+        for(int j = 0; j < n_snap-1; ++j)
+        {
+            h_temp = snapshot_ts[j+1]->getData()[0] - snapshot_ts[j]->getData()[0];
+            h.push_back(h_temp);
+            delta_temp = (snapshots[j+1]->getData()[i] - snapshots[j]->getData()[i])/h_temp;
+            delta.push_back(delta_temp);
+            t_in.push_back(snapshot_ts[j]->getData()[0]);
+            y_in.push_back(snapshots[j]->getData()[i]);
+        }
+        t_in.push_back(snapshot_ts[n_snap-1]->getData()[0]);
+        y_in.push_back(snapshots[n_snap-1]->getData()[i]);
+
+        double d_temp = ((2*h[0] + h[1])*delta[0] - h[0]*delta[1])/(h[0]+h[1]);
+        if(sign(d_temp)!=sign(delta[0]))
+        {
+            d_temp = 0;
+        }
+        else if (sign(delta[0]) != sign(delta[1]) && abs(d_temp) > abs(3*delta[0]))
+        {
+            d_temp = 3*delta[0];
+        }
+        d.push_back(d_temp);
+        int counter = 0;
+        for(int j = 0; j < n_snap-2; ++j)
+        {
+            d_temp = computeDerivative(delta[j],delta[j+1],h[j],h[j+1]);
+            d.push_back(d_temp);
+            while(output_ts[counter]->getData()[0] <= t_in[j+1])
+            {
+                t = output_ts[counter]->getData()[0];
+                output_snapshots[counter]->getData()[i] = y_in[j]*computeH1(t,t_in[j],
+                        t_in[j+1]) +
+                        y_in[j+1]*computeH2(t,t_in[j],t_in[j+1]) +
+                        d[j]*computeH3(t,t_in[j],t_in[j+1]) +
+                        d[j+1]*computeH4(t,t_in[j],t_in[j+1]);
+                counter += 1;
+            }
+        }
+
+        d_temp = ((2*h[n_snap-2]+h[n_snap-3])*delta[n_snap-2] - h[n_snap-2]*delta[n_snap
+                  -3])/(h[n_snap-2]+h[n_snap-3]);
+        if(sign(d_temp) != sign(delta[n_snap-2]))
+        {
+            d_temp = 0;
+        }
+        else if (sign(delta[n_snap-2]) != sign(delta[n_snap-3])
+                 && abs(d_temp) > abs(3*delta[n_snap-2]))
+        {
+            d_temp = 3*delta[n_snap-2];
+        }
+        d.push_back(d_temp);
+
+        while(counter < n_out
```

Comment:
Is it guaranteed that `counter == n_out` at the very end of the function, so that `output_snapshots` is fully defined, or could there be undefined data at the end in some cases? Should there be a `CAROM_VERIFY` for the condition `counter == n_out` at the end?

Reply:
This is the intended behavior.  I added the suggested `CAROM_VERIFY` statement.